KAFKA-17439: Make polling for new records an explicit action/event in the new consumer

kirktrue opened the pull request:
Updated the `FetchRequestManager` to only create and enqueue fetch requests when signaled to do so by a `FetchEvent`.

The application thread and the background thread each contains logic that is performed if there is buffered data from a previous fetch. There's a race condition because the presence of buffered data could change between the two threads’ respective checks. Right now the window for the race condition to occur is _wide open_; this change aims to make the window _ajar_.

In the `ClassicKafkaConsumer`, the application thread will explicitly issue fetch requests (via the `Fetcher` class) at specific points in the `Consumer.poll()` cycle. Prior to this change, the `AsyncKafkaConsumer` would issue fetch requests independently from the user calling `Consumer.poll()`; the fetches would happen nearly continuously as soon as _any_ assigned partition was fetchable. With this change, the `AsyncKafkaConsumer` introduces a `FetchEvent` that signals to the background thread that a fetch request should be issued. The specific points where this is done in the `Consumer.poll()` cycle of the `AsyncKafkaConsumer` now match the `ClassicKafkaConsumer`. In short: this makes `AsyncKafkaConsumer` act nearly identical to the `ClassicKafkaConsumer` in this regard.

As mentioned above, this change does not completely solve the problem related to fetch session eviction. Exactly how the window where the race condition can be shut completely is outside the scope of this change.

See KAFKA-17182.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

kirktrue committed:
KAFKA-17439: Make polling for new records an explicit action/event in the new consumer

Updated the FetchRequestManager to only create and enqueue fetch requests when signaled to do so by a FetchEvent.

StanleyWu8787 commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -683,10 +683,6 @@ public ConsumerRecords<K, V> poll(final Duration timeout) {
-                // Make sure to let the background thread know that we are still polling.
-                applicationEventHandler.add(new PollEvent(timer.currentTimeMs()));
-
                 // We must not allow wake-ups between polling for fetches and returning the records.
                 // If the polled fetches are not empty the consumed position has already been updated in the polling
                 // of the fetches. A wakeup between returned fetches and returning records would lead to never
```
Just curious.
Do we still need `wakeupTrigger.maybeTriggerWakeup()` below?

kirktrue replied:
That's a great question, @StanleyWu8787! I will admit I don't really understand that aspect of the code, which is why I haven't answered your question more satisfactorily yet. I will take a look.

kirktrue replied:
This may be moot, given I've reverted how this works.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/ApplicationEventProcessor.java:
```diff
@@ -159,6 +159,9 @@ private void process(final PollEvent event) {
             requestManagers.shareHeartbeatRequestManager.ifPresent(hrm -> hrm.resetPollTimer(event.pollTimeMs()));
         }
+
+        CompletableFuture<Void> future = requestManagers.fetchRequestManager.requestFetch();
```
Hijacking the poll event like this is not quite going to work for the share consumer. In the share consumer, we use the PollEvent to signal that polling is occurring, and we have a separate ShareFetch event that asks the ShareConsumeRequestManager to fetch more records. There is no FetchRequestManager in a share consumer.

kirktrue replied:
I take umbrage at your use of "hijack." I was simply _borrowing_ it. 😏

Regardless, I've introduced a separate `FetchEvent` that should disentangle the two tasks.

AndrewJSchofield replied:
Thank you. Much nicer.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. I'm quite surprised to see how little code needed to change to make the fetching explicit. A couple of points.

First, I think it better not to overload the PollEvent because that's already used in the share consumer.

Second, it seems to me that there is still the potential for over-fetching, and this will still cause churn of the fetch session cache.

In the case where the consumer is only fetching a single partition, I think it works pretty well. The set of fetchable partitions will be empty if there's buffered data, and contain the only partition in the fetch session if there is not. So, you'll only send a Fetch request when there's a need for more data and the fetch session will not churn.

In the case where the consumer is fetching more than one partition on a particular node, if a subset of the partitions is fetchable, then the fetch session will be modified by sending a Fetch request and that seems to have the potential for a lot of churn.

Of course, all of this code is in common between the legacy consumer and the async consumer. The async consumer is still very keen on fetching so I don't properly grasp why this PR would make the fetch session behaviour better.

kirktrue commented:
Hi @AndrewJSchofield!

Thanks for the review 👍 

> First, I think it better not to overload the PollEvent because that's already used in the share consumer.

Agreed. I've introduced a `FetchEvent` so that the two separate mechanisms won't step on each others' toes.

> Second, it seems to me that there is still the potential for over-fetching, and this will still cause churn of the fetch session cache.

Agreed. It aims to _lessen_ the churn. _Preventing_ the churn completely is a future task. 

> In the case where the consumer is only fetching a single partition, I think it works pretty well. The set of fetchable partitions will be empty if there's buffered data, and contain the only partition in the fetch session if there is not. So, you'll only send a Fetch request when there's a need for more data and the fetch session will not churn.

Correct.

> In the case where the consumer is fetching more than one partition on a particular node, if a subset of the partitions is fetchable, then the fetch session will be modified by sending a Fetch request and that seems to have the potential for a lot of churn.

Correct again!

Any partition with buffered data at the point where the fetch request is being generated will be marked as "removed" from the broker's fetch session cache. That's the crux of the problem 😞

Something that I tend to lose sight of is the fact that it's not a foregone conclusion that a fetch session will be evicted when it has partitions removed. Of course, it will increase its eligibility for eviction if the broker hosting the fetch session is resource-constrained and invokes the eviction process.

> Of course, all of this code is in common between the legacy consumer and the async consumer.

I'm not sure I follow. This code is all specific to the `AsyncKafkaConsumer`. While the `ClassicKafkaConsumer` has a similar race condition, it is 2-4 orders of magnitude less likely to happen.

> The async consumer is still very keen on fetching so I don't properly grasp why this PR would make the fetch session behaviour better.

Yep—the design of the `AsyncKafkaConsumer` fetching continuously in the background makes it _very_ keen to cause this problem. With this change, the application thread now signals when to fetch, which results in the background thread creating and issuing the fetch requests _much_ less often.

Thanks!

kirktrue commented:
@AndrewJSchofield, et al.—it can be helpful to compare the flow of `ClassicKafkaConsumer.poll()` and `AsyncKafkaConsumer.poll()`, specifically how it invokes fetch. Note that the `sendFetches()` method name, as well as when it is invoked, comes from the `ClassicKafkaConsumer.poll()`. So this is really making the new consumer act much more like the old one.

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/FetchEvent.java:
```diff
@@ -0,0 +1,36 @@
+ *
+ * <em>Note</em>: this event is completed when the {@link FetchRequestManager} has finished performing the
+ * fetch request process. It does not mean that the requests are complete. It could be the case that no fetch
+ * requests were created. Also of note is that if any fetch requests were created.
```
I don't quite understand the sentence "Also of note is that if any fetch requests were created.".  I think I understand what the PR does - this event completes when the background thread has created any FetchRequests to be sent by the network client delegate. There might be none to send and the event completes in this case too.

kirktrue replied:
I updated the comment to be slightly less awkward. LMK if it still needs work. Thanks!

AndrewJSchofield commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchRequestManager.java:
```diff
@@ -65,16 +67,59 @@ protected void maybeThrowAuthFailure(Node node) {
+            // In this case, we have an outstanding fetch request, so chain the newly created future to be
+            // invoked when the outstanding fetch request is completed.
+            pendingFetchRequestFuture.whenComplete((value, exception) -> {
+                if (exception != null) {
```
Given that you only complete the pending event successfully in the finally block of `poll()`, you could probably just `future.complete(null)`.

kirktrue replied:
Just to make sure I understand: are you stating that because `pendingFetchRequestFuture` is only ever `complete()`-ed and not `completeExceptionally()`-ed, that the code in `requestFetch()` could be simplified to:

```java
if (pendingFetchRequestFuture != null) {
    // In this case, we have an outstanding fetch request, so chain the newly created
    // future to be completed when the outstanding fetch request is completed.
    pendingFetchRequestFuture.whenComplete((__, ___) -> future.complete(null));
} else {
    pendingFetchRequestFuture = future;
}
```

Or am I misunderstanding?

Thanks!

AndrewJSchofield replied:
Yes.

kirktrue replied:
FYI: I updated the code to catch spurious errors from `prepareFetchRequests()` that could result in needing to complete the chained `CompletableFuture` with an error.

AndrewJSchofield reviewed (changes requested):
Thanks for the updates and for the explanation of the mechanism.

I think it would be appropriate to test in `FetchRequestManagerTest` the mechanism of the pending fetch request future in the various permutations.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchRequestManager.java:
```diff
@@ -65,16 +67,59 @@ protected void maybeThrowAuthFailure(Node node) {
-        return pollInternal(
+        if (pendingFetchRequestFuture == null) {
+            // If no explicit request for fetching has been issued, just short-circuit.
+            return PollResult.EMPTY;
```
this fix is huge (positive). Regardless of the session eviction situation, we had a `FetchRequestManager` (continuously polled in the background as soon as the new consumer is created) that could generate fetch requests without a call to consumer.poll in the app thread, is my understanding correct?

kirktrue replied:
Correct 😄

We viewed the continuous background fetch as a net positive (and maybe it is), but for now we're trying to reduce the difference in when and how the two consumers request messages from the cluster.

Thanks!

kirktrue committed:
Merge remote-tracking branch 'origin/trunk' into KAFKA-17439-poll-explicitly

kirktrue committed:
Minor tweaks to FetchEvent documentation.

kirktrue committed:
Update FetchRequestManager.java

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchRequestManager.java:
```diff
@@ -65,16 +67,59 @@ protected void maybeThrowAuthFailure(Node node) {
+     * <p/>
+     *
+     * The returned {@link CompletableFuture} is {@link CompletableFuture#complete(Object) completed} when the
+     * fetch request(s) have been created and enqueued into the network client's outgoing send buffer.
```
To be accurate, it completes when the fetch request has been generated, not yet enqueued really (because it completes on the finally clause on the manager.poll, and the generated request will be enqueued later on, after all managers are polled), is my understanding correct?

kirktrue replied:
Wow, I finally met someone even more pedantic than me 😏

You're correct, I'll update the docs. 😄

kirktrue replied:
Updated the comments. Please take a look and let me know if it needs more massaging. Thanks!

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchRequestManager.java:
```diff
@@ -65,16 +67,59 @@ protected void maybeThrowAuthFailure(Node node) {
+     * fetch request(s) have been created and enqueued into the network client's outgoing send buffer.
+     * It is <em>not completed</em> when the network client has received the data.
+     *
+     * @return Future for which the caller can wait to ensure that the requests have been enqueued
```
as above? the completion only tells that the request has been generated right?

kirktrue replied:
Updated the comments. Thanks!

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchRequestManager.java:
```diff
@@ -65,16 +67,59 @@ protected void maybeThrowAuthFailure(Node node) {
-        );
+            );
+        } finally {
+            // Completing the future here means that the caller knows that the fetch request logic has been
```
"the fetch request logic has been performed" seems confusingly vague (could hint more that what we really do here I'm afraid). Maybe just saying that the caller knows that the fetch request has been generated?

kirktrue replied:
Tweaked this text too. Take a look. Thanks!

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -1709,6 +1715,10 @@ private void updateLastSeenEpochIfNewer(TopicPartition topicPartition, OffsetAnd
     }
 
+    private void sendFetches(Timer timer) {
+        applicationEventHandler.addAndGet(new FetchEvent(calculateDeadlineMs(timer)));
```
this new event is really about `SendFetch` (not actual fetching), so maybe it makes the intention clearer if we rename it, aligned with the sendFetches func that creates it? Maybe me but reading FetchEvent here makes me go to the AppEventProcessor to make sure that we're not fetching/blocking there :) and it's only for sending the request

kirktrue replied:
I started off renaming it as `SendFetch` but then that didn't seem right because the event doesn't directly send anything, either. Then I named it `GenerateFetchRequestsEvents`, which is... not good. I settled on `CreateFetchRequestsEvents` because that's exactly what it does.

I am not _too_ worried about the name, so if you really prefer `SendFetch` I can keep it as that.

Thanks.

kirktrue committed:
Added unit tests to exercise poll, request-then-poll, and duplicate requests

kirktrue committed:
Updated FetchEvent to CreateFetchRequestsEvent and catching errors from prepareFetchRequests()

kirktrue committed:
Fixed spacing issue that checkstyle wasn't happy with

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchRequestManager.java:
```diff
@@ -65,16 +69,60 @@ protected void maybeThrowAuthFailure(Node node) {
+            // A "dummy" poll result is returned here rather than rethrowing the error because any error
+            // that is thrown from any RequestManager.poll() method interrupts the polling of the other
+            // request managers.
+            pendingFetchRequestFuture.completeExceptionally(t);
```
Even though we propagate the error here, we just ignore it at the consumer level when doing addAndGet on this result. Should we at least add an error log for visibility? (there is some usage of the subscription state when handling the success/failure, so if we get some unexpected error from that, like the no current assignment, it would go totally silent it seems)

kirktrue replied:
In `sendFetches()` the error from the background thread will be thrown, so it would bubble up to the user.

I'm prone to code blindness 😄 Where is the result of `addAndGet()` being ignored?

kirktrue replied:
Sorry to bug you on this @lianetm...

I looked again at the code path in `sendFetches()` which—in the case of blocking fetch—will throw the exception if the `Future` is completed exceptionally. Can you point out where the exception is being suppressed?

Thanks!

lianetm replied:
you're right, I see the error propagated via the chained futures so all good, I truly don't quite get/remember what I had in mind with this at that time sorry, no concerns now.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java:
```diff
@@ -3387,6 +3394,71 @@ public void testWhenFetchResponseReturnsALeaderShipChangeErrorAndNewLeaderInform
+        assertFalse(future.isDone());
+
+        assertDoesNotThrow(() -> sendFetches(false));
+        assertTrue(future.isCompletedExceptionally());
```
I guess here we expect the same auth exception? if so we could better `assertFutureThrows`

kirktrue replied:
Good suggestion!

kirktrue replied:
Done. PTAL. Thanks!

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -697,6 +698,8 @@ public ConsumerRecords<K, V> poll(final Duration timeout) {
                 updateAssignmentMetadataIfNeeded(timer);
                 final Fetch<K, V> fetch = pollForFetches(timer);
                 if (!fetch.isEmpty()) {
+                    sendFetches(timer);
```
this could throw `TimeoutException` I expect (poll with zero/low timeout + addAndGet) right? before this change, not having enough time to fetch would just carry on, exit the while(timer.notExpired) and return empty records. I image we could catch the timeout that the addAndGet could throw, probably log, and just carry on (to align with the previous behaviour)? (to review on the other places where we call the sendFEtches too)

kirktrue replied:
You're right. I think this should submit the request via `add()` (vs. `addAndGet`) in the prefetch case.

kirktrue replied:
`sendFetches()` now takes a flag to determine if it should wait:

```java
    private void sendFetches(Timer timer, boolean shouldWait) {
        CreateFetchRequestsEvent event = new CreateFetchRequestsEvent(calculateDeadlineMs(timer));

        if (shouldWait) {
            applicationEventHandler.addAndGet(event);
        } else {
            applicationEventHandler.add(event);
        }
    }
```

This allows prefetch to submit the request for fetch without waiting and the blocking fetch in `pollForFetches()` to wait for it to be sent out.

Does that seem like a sufficient solution?

Thanks!

lianetm reviewed (commented):
Hey @kirktrue , thanks for the updates, some comments...

kirktrue committed:
Merge branch 'trunk' into KAFKA-17439-poll-explicitly

kirktrue committed:
Update KafkaConsumerTest.java

kirktrue committed:
Merge branch 'trunk' into KAFKA-17439-poll-explicitly

kirktrue committed:
Update AsyncKafkaConsumer so that prefetch requests don't block

kirktrue committed:
Including fix for test that uses add() instead of addAndGet()

kirktrue committed:
Added requestGenerated back that somehow was removed

kirktrue committed:
No longer need to wait for FETCH RPC in test since fetches don't happen continuously anymore

kirktrue committed:
Testing that Future.get() throws a specific exception type

kirktrue committed:
Fixed spotless complaints

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -1531,6 +1534,9 @@ private Fetch<K, V> pollForFetches(Timer timer) {
         }
 
+        // send any new fetches (won't resend pending fetches)
+        sendFetches(timer, true);
```
Similar concern as in the other call to sendFetches. I wonder if this 'true' here could be dangerous in cases of poll(0/lowTimeout)? In those cases, we could easily get a `TimeoutException` in this line because of the `addAndGet`, and we would break out of here without making it to awaitNotEmpty right? I guess that the overall situation would be sorted out on the next poll iteration, that would find the fetch results available, but still doesn't feel right to break on Timeout on the "generate request" phase, instead of on the "awaitNotEmpty" phase.

At this point I'm only spotting a difference that doesn't feel conceptually right, but honestly not sure if the implications are truly harmful in this case. The alternative I see is to pass shouldWait here false to ensure we don't block when telling the background to generate fetches, then the awaitNotEmpty is the one to apply the timer. What do you think?

kirktrue replied:
True. `ClassicKafkaConsumer.sendFetches()` does not make use of a `Timer` or timeout of any sort, so the act of creating the outgoing requests shouldn't yield any timeouts.

`AsyncKafkaConsumer.sendFetches()` does take a `Timer`, only because we do want to wait for the requests to be created. And the easiest way to wait for the requests is to use the `addAndGet()` approach. If we were to switch to `add()`, we wouldn't _technically_ be waiting for the requests to be created, but I'm not sure that's really that much of a concern in practice.

So in the case you mentioned (small timeout values), if we send the `CreateFetchRequestsEvent` but don't wait for it, then we shift the possibility of the `TimeoutException` to the `FetchBuffer.awaitNotEmpty()` call. I don't think that's a problem.

So, correct me if I'm wrong, but it sounds like we want to change `CreateFetchRequestsEvent` so that it isn't a `CompletableApplicationEvent` since we don't want to wait in either case of fetch requests: blocking or prefetch. I'm fine with that and it'll make the code a little cleaner too.

kirktrue replied:
So I realized I couldn't (comfortably) remove the wait because one of the byproducts is the ability to catch errors and bubble them up. PTAL at the latest (with probably too many comments) to see what you think. Thanks!

lianetm reviewed (commented):
Thanks for the updates!

kirktrue committed:
Merge branch 'trunk' into KAFKA-17439-poll-explicitly

kirktrue committed:
Update AsyncKafkaConsumer.java

Allowing sendFetches to block in order to communicate errors, but it will suppress timeouts.

kirktrue committed:
Merge branch 'trunk' into KAFKA-17439-poll-explicitly

kirktrue committed:
Reducing diff noise

kirktrue committed:
Updated test to check that CreateFetchRequests event uses addAndGet

kirktrue commented:
@lianetm—tests are passing and all comments have been addressed. Can you make another review pass? Thanks!

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchRequestManager.java:
```diff
@@ -65,16 +69,60 @@ protected void maybeThrowAuthFailure(Node node) {
                 this::handleFetchFailure
-        );
+            );
+            pendingFetchRequestFuture.complete(null);
```
do we need to complete this future also on `pollOnClose`? there may be a `pendingFetchRequestFuture` there that won't be completed (not that I'm seeing how leaving that future uncompleted on close will cause a problem but seems safer to complete it, consistently with how we do it here after `pollInternal`)

kirktrue replied:
I've moved the `Future`-handling code to `pollInternal()` for consistency. LMK what you think.

lianetm commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetchRequestManagerTest.java:
```diff
@@ -3386,6 +3394,71 @@ public void testWhenFetchResponseReturnsALeaderShipChangeErrorAndNewLeaderInform
+    public void testPollWithoutCreateFetchRequests() {
+        buildFetcher();
+
+        assignFromUser(singleton(tp0));
+        subscriptions.seek(tp0, 0);
```
what do we need these for? This test is about polling the fetchReqMgr without having created requests, which should simply notice it has no pendingFetch and return empty poll result (so I would expect we don't need to setup anything regarding subscriptions)

kirktrue replied:
Prior to this change, all that was needed to start fetching was a valid set of assigned partitions. Post-change, that is insufficient; no fetch requests will be issued if not explicitly requested.

If I removed the assignment, the test looks very anemic:

```java
    @Test
    public void testPollWithoutCreateFetchRequests() {
        buildFetcher();
        assertEquals(0, sendFetches(false));
    }
```

I'm OK to remove that test since the other tests cover that basic test anyway.

Thoughts?

lianetm replied:
thinking more I say it's better to keep them even if they are setting up a condition that it's truly not used in the test. The reason is really that if ever there's a regression and we wrongfully send a "ghost" request in the sendFetches(false), we don't want to miss catching it just because some other conditions are missing.

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -1520,6 +1523,9 @@ private Fetch<K, V> pollForFetches(Timer timer) {
         }
 
+        // send any new fetches (won't resend pending fetches)
+        sendFetches(timer);
```
The actual poll now happens in here (addAndGet that will complete when the background has had one run, called fetchMgr.poll), so should the log line on ln 1538 "Polling for fetches with timeout..." be right before this?

kirktrue replied:
We're not polling for the incoming responses in `sendFetches()`, just enqueuing the outgoing requests. This mimics the `ClassicKafkaConsumer` in that the requests are enqueued in its `sendFetches()` but then toward the bottom of `pollForFetches()` `client.poll()` is invoked to wait for the results of the fetch requests.

lianetm replied:
well the `sendFetches` blocks until the `CreateFetchRequestsEvent` completes, and that only happens on fetchMgr.poll
https://github.com/apache/kafka/blob/aae7e97fa041dac34a0788f90e837ff7f5ec3846/clients/src/main/java/org/apache/kafka/clients/consumer/internals/FetchRequestManager.java#L115
So when the `sendFetches` completes we did poll the manager right? (and depending on time, maybe we did poll the client.poll too, which happens in the background right after polling all managers). That's why the log for "Polling for fetches" made sense to me before the `sendFetches`, but am I missing another poll happening after the log line maybe? (where it is now)

kirktrue replied:
The two `ConsumerDelegate` implementations work differently:

* `AsyncKafkaConsumer`: `FetchRequestManager.poll()` will complete the event's `Future` on the background thread before it exits, i.e. before the thread starts the network I/O. Completing the `Future` starts the application thread racing toward logging that message and the background thread racing toward starting network I/O. I'll admit—I haven't dug through the code to surmise the relative costs of each thread's work before either cross their respective finish lines to win.
* `ClassicKafkaConsumer`: `Fetcher.sendFetchesInternal()` calls `ConsumerNetworkClient.send()` to enqueue the request, but then it calls `NetworkClient.wakeup()`. Since the same `ConsumerNetworkClient` instance used by the consumer is also used by `AbstractCoordinator.HeartbeatThread`, it's technically possible that the heartbeat thread's `run()` method could start network I/O when it calls `NetworkClient.pollNoWakeup()`. Granted, that's a race that the application thread is much more likely to win given that the heartbeat thread runs much less frequently.

Here are some points to consider:

* The definition of the term "poll" as used in the log is open to interpretation. The term "poll" is everywhere, making its meaning ambiguous at any given point of use 😢
* I agree there is a race condition (for both consumers, but more likely for the new consumer) that could result the the log message being emitted after the network I/O has commenced
* For this to pose a problem to users, there needs to be other log entries that we're racing with, right?. We're trying to avoid the condition where the user is confused/mislead because the entries in the log are emitted in non-deterministic ordering.
* The log line in question is only output at level `TRACE`, which I _assume_ is very rare for users to enable.

Given the above, I'm of the opinion that it's an exercise in hair splitting to alter the logging. However, I could also just change it which would have been way less effort than researching, thinking, and composing this response 🤣

If we leave the log line as it is, what would the effect be for the user?

lianetm replied:
I surely didn't intend for you to put up that long response he he, sorry. It's not about the log line per-se, it's about the alignment on where the poll happens. The classic consumer logs "Polling for records", then calls `client.poll`. vs us here we do `sendFetches` (which triggers the client.poll async in the background thread because it blocks until we poll the fetch manager), then log "Polling for fetches...". 

That's the diff I saw and just wanted to understand/align on where the poll happens: once we trigger `sendFetches` (blocking), the client.poll will happen in the background anytime, not controlled by the app thread. Agreed? If so I'm ok with leaving the log unchanged, understanding it could come out after the client.poll happened.

kirktrue replied:
> That's the diff I saw and just wanted to understand/align on where the poll happens: once we trigger sendFetches (blocking), the client.poll will happen in the background anytime, not controlled by the app thread. Agreed?

Agreed—the background thread is going to move from calling each of the `RequestManager`’s `poll()` method to `NetworkClient.poll()` method without the intervention of the application thread.

> If so I'm ok with leaving the log unchanged, understanding it could come out after the client.poll happened.

Thanks!

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -707,6 +708,8 @@ public ConsumerRecords<K, V> poll(final Duration timeout) {
                 updateAssignmentMetadataIfNeeded(timer);
                 final Fetch<K, V> fetch = pollForFetches(timer);
                 if (!fetch.isEmpty()) {
+                    sendFetches(timer);
```
at this point we may already have records in hand to return (consumed position updated), so we should be very careful to not throw any error here. But this `sendFetches` could throw interrupted because of the `addAndGet` right? 

Shouldn't we just do a best effort to pipeline the next requests using `add` instead of `addAndGet`? It would achieve what we want, removing the risk of errors, and it would actually align better with what the classic does on this `sendFetches` + `transmitSends`:
https://github.com/apache/kafka/blob/140d35c5459f4c7a91b23a238467cb5aa01b59fb/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java#L327-L329

kirktrue replied:
Done.

lianetm reviewed (commented):
Thanks for the updates @kirktrue! Took another pass and left some comments for consideration.

kirktrue committed:
Merge branch 'trunk' into KAFKA-17439-poll-explicitly

kirktrue committed:
Updates to suppress exceptions for pre-fetch and handle pollOnClose()

kirktrue committed:
Updates to let pre-fetch be asynchronous

kirktrue committed:
Tweak to comment

lianetm commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java:
```diff
@@ -1606,6 +1618,73 @@ private void updateLastSeenEpochIfNewer(TopicPartition topicPartition, OffsetAnd
+     *
+     * <ul>
+     *     <li>
+     *         The method will wait for confirmation of the request creation before continuing.
```
This is not true now for prefetching that uses `.add` instead of `.addAndGet`, should we remove this line?

kirktrue replied:
Good catch. Reworded to state that it will not wait for confirmation.

lianetm reviewed (commented):
Thanks for the updates @kirktrue! Just one nit left, almost there.

kirktrue committed:
Merge branch 'trunk' into KAFKA-17439-poll-explicitly

kirktrue committed:
Updated comments for sendPrefetches to correctly reflect implementation

lianetm reviewed (approved):
Thanks for all the updates @kirktrue! LGTM.
